### PR TITLE
Leaving out params crashes tm rpc

### PR DESCRIPTION
**Pull request by melekes**
See #708

### CODE REVIEW

**Comment by melekes** [comment]
> Could we also ensure that that's the only path to those functions?

what do you mean? try with and without "params: {}"? I already did that

**Comment by odeke-em** [comment]
By path I meant, like code usage so perhaps a grep for usage of those functions might suffice?

File: rpc/core/types/responses.go [hunk 1]
```diff
@@ -75,8 +76,8 @@ type ResultValidators struct {
 }
 
 type ResultDumpConsensusState struct {
-	RoundState      string   `json:"round_state"`
-	PeerRoundStates []string `json:"peer_round_states"`
+	RoundState      json.RawMessage   `json:"round_state"`
```
Comment (ebuchman) [inline rpc/core/types/responses.go]: Oh - does this make it easier to parse the result?!

**Comment by ebuchman** [comment]
CI failure:

```
Step 13 : RUN go install ./cmd/tendermint
 ---> Running in c30821f47119
node/node.go:26:2: 
rpc/lib/doc.go:102:4: expected 'package', found 'EOF'
```

File: rpc/core/types/responses.go (hunk 1 above)
Comment (melekes) [reply rpc/core/types/responses.go]: yeah.. before we're spitting a string (2 strings), now it's a JSON https://github.com/tendermint/tendermint/issues/708#issuecomment-335115500

**Comment by melekes** [comment]
I fixed the tests in https://github.com/tendermint/tendermint/pull/721 https://github.com/tendermint/tendermint/pull/721/commits/6fca03c36af1f0ab20434f02a8a84b9879d98fe1

File: rpc/core/types/responses.go (hunk 1 above)
Comment (ebuchman) [reply rpc/core/types/responses.go]: So, it seems like we could use the actual structs here, but we don't want to have to import consensus to get them. So probably we should move some types from `consensus` into `consensus/types` so we can import here. 

Will these raw messages be identical to:

```
type ResultDumpConsensusState struct {
  RoundState cstypes.RoundState
  PeerRoundStates map[string]cstypes.PeerRoundState
}
```

**Comment by ebuchman** [comment]
This also causes a panic in the non-ws RPC, but that server's handler is protected by a `recover` statement, so you only see the panic in the logs.

Is there a reason the RawMessage is a pointer in the first place?

**Comment by ebuchman** [comment]
We should include this in 0.11.1. The format of dump_consensus was so poor previously it shouldn't have been used - a great bug fix

**Comment by odeke-em** [comment]
Interestingly the fuzzer finds the other crash in 9s
given code
```go
package handler

import (
	"bytes"
	"encoding/json"
	"io/ioutil"
	"net/http"
	"net/http/httptest"

	rs "github.com/tendermint/tendermint/rpc/lib/server"
	types "github.com/tendermint/tendermint/rpc/lib/types"
	"github.com/tendermint/tmlibs/log"
)

var rpcFuncMap = map[string]*rs.RPCFunc{
	"c": rs.NewRPCFunc(func(s string, i int) (string, int) { return "foo", 200 }, "s,i"),
}

func Fuzz(data []byte) int {
	mux := http.NewServeMux()
	buf := new(bytes.Buffer)
	lgr := log.NewTMLogger(buf)
	rs.RegisterRPCFuncs(mux, rpcFuncMap, lgr)

	req, _ := http.NewRequest("POST", "http://localhost/", bytes.NewReader(data))
	rec := httptest.NewRecorder()
	mux.ServeHTTP(rec, req)
	res := rec.Result()
	blob, err := ioutil.ReadAll(res.Body)
	if err != nil {
		panic(err)
	}
	recv := new(types.RPCResponse)
	if err := json.Unmarshal(blob, recv); err != nil {
		panic(err)
	}
	return 1
}
```

We could also adapt the code in the fuzzer for a test.

**Comment by odeke-em** [comment]
where fuzzer crash output is
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x133a066]

goroutine 1 [running]:
github.com/tendermint/tendermint/rpc/lib/server.jsonParamsToArgsRPC(0xc4201b32d0, 0x0, 0xc4201f2d58, 0x1, 0xc42007b198, 0x0, 0x0)
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/gopath/src/github.com/tendermint/tendermint/rpc/lib/server/handlers.go:213 +0x26
github.com/tendermint/tendermint/rpc/lib/server.makeJSONRPCHandler.func1(0x14253a0, 0xc4201e15c0, 0xc420205b00)
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/gopath/src/github.com/tendermint/tendermint/rpc/lib/server/handlers.go:130 +0x3b7
net/http.HandlerFunc.ServeHTTP(0xc42023a480, 0x14253a0, 0xc4201e15c0, 0xc420205b00)
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/goroot/src/net/http/server.go:1921 +0x44
net/http.(*ServeMux).ServeHTTP(0xc4202299e0, 0x14253a0, 0xc4201e15c0, 0xc420205b00)
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/goroot/src/net/http/server.go:2299 +0x130
github.com/tendermint/tendermint/fuzz/rpc/lib/server/handler.Fuzz(0x2230000, 0x17, 0x200000, 0x3)
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/gopath/src/github.com/tendermint/tendermint/fuzz/rpc/lib/server/handler/main.go:27 +0x1f4
go-fuzz-dep.Main(0x13fc5e8)
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/goroot/src/go-fuzz-dep/main.go:49 +0xad
main.main()
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/gopath/src/github.com/tendermint/tendermint/fuzz/rpc/lib/server/handler/go.fuzz.main/main.go:10 +0x2d

goroutine 5 [syscall]:
os/signal.signal_recv(0x0)
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/goroot/src/runtime/sigqueue.go:131 +0xa7
os/signal.loop()
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/goroot/src/os/signal/signal_unix.go:22 +0x52
created by os/signal.init.0
	/var/folders/v3/7z434qpx5v3bw0wh8h2myfpw0000gn/T/go-fuzz-build199365166/goroot/src/os/signal/signal_unix.go:28 +0x59
exit status 2
```
and the other unsafe spot is 
https://github.com/tendermint/tendermint/blob/bcca27ee2032299a7c571ae62a87b596eb96a899/rpc/lib/server/handlers.go#L130

and the input is
```JSON
{"method":"c","id":"0"}
```

**Comment by odeke-em** [comment]
I actually think that just like @ebuchman might have alluded to in https://github.com/tendermint/tendermint/pull/724#issuecomment-335316484, if we don't make Params a pointer, and remove the extra indirection, we won't even need the current PR.
json.RawMessage is an alias for `[]byte` and we can always just check if it wasn't set with a
```go
if len(Params) == 0 {
   return
}
```

File: rpc/core/types/responses.go (hunk 1 above)
Comment (melekes) [reply rpc/core/types/responses.go]: Good ideas! Will do now

Comment (melekes) [reply rpc/core/types/responses.go]: > clients are importing too much crap

that was the reason behind `json.RawMsg`

Comment (melekes) [reply rpc/core/types/responses.go]: moving types will be a breaking change, though

Comment (melekes) [reply rpc/core/types/responses.go]: see https://github.com/tendermint/tendermint/pull/724/commits/dc0e8de9b0671fc93d7479b20c2c9b5ea54c48c6. I did not move all of the types. just ones which are necessary

**Comment by melekes** [comment]
@odeke-em @ebuchman addressed all your comments. PTAL

**Comment by melekes** [comment]
@odeke-em thanks for finding another bug! fixed it too.

File: rpc/core/consensus.go [hunk 2]
```diff
@@ -82,14 +82,12 @@ func Validators(heightPtr *int) (*ctypes.ResultValidators, error) {
 // }
 // ```
 func DumpConsensusState() (*ctypes.ResultDumpConsensusState, error) {
-	roundState := consensusState.GetRoundState()
-	peerRoundStates := []string{}
+	peerRoundStates := make(map[string]*cstypes.PeerRoundState)
 	for _, peer := range p2pSwitch.Peers().List() {
 		// TODO: clean this up?
```
Comment (ebuchman) [inline rpc/core/consensus.go]: think we can remove this comment now!

File: rpc/lib/server/handlers.go [hunk 3]
```diff
@@ -127,10 +129,13 @@ func makeJSONRPCHandler(funcMap map[string]*RPCFunc, logger log.Logger) http.Han
 			WriteRPCResponseHTTP(w, types.RPCMethodNotFoundError(request.ID))
 			return
 		}
-		args, err := jsonParamsToArgsRPC(rpcFunc, request.Params)
-		if err != nil {
-			WriteRPCResponseHTTP(w, types.RPCInvalidParamsError(request.ID, errors.Wrap(err, "Error converting json params to arguments")))
-			return
+		var args []reflect.Value
+		if len(request.Params) > 0 {
```
Comment (ebuchman) [inline rpc/lib/server/handlers.go]: as far as I can tell, now that params is not a pointer, we shouldn't need this guard, but it's fine to leave it

Comment (melekes) [reply rpc/lib/server/handlers.go]: no, we need it because overwise this results in the following error: "params should be a map or array"

**Pull request merged**